File: Semester3/Advanced Programming Methods/Labs/lab3/src/domain/validators/GradeValidator.java
package domain.validators;

import domain.entities.Grade;

import java.util.function.Predicate;

public class GradeValidator implements Validator<Grade> {

    /**
     * Overriding validate function for grade entity
     * @param entity - grade to be validated
     * @throws ValidatorException - if the grade is incorrect
     * @throws IllegalArgumentException - if the arguments are invalid
     */
    @Override
    public void validate(Grade entity) throws ValidatorException, IllegalArgumentException {
        if(entity == null){
            throw new IllegalArgumentException("Grade can't be null!");
        }
        String errs = "";

        Predicate<Grade> emptyProfName = (x) -> x.getProfName().equals("");
        Predicate<Grade> negativeGrade = (x) -> x.getGrade() < 0;
        Predicate<Grade> tooHighGrade = (x) -> x.getGrade() > 10;
        Predicate<Grade> negativeStId = (x) -> x.getStudId() <= 0;
        Predicate<Grade> negativeHwId = (x) -> x.getHwId() <= 0;
        Predicate<Grade> emptyFeedBack = (x) -> x.getFeedback().equals("");
        Predicate<String> emptyErrs = x -> x.equals("");

        if(emptyProfName.test(entity)){
            errs += "Professor's name can't be empty!\n";
        }
        if(negativeGrade.test(entity)){
            errs += "Grade must be a positive integer!\n";
        }
        if(tooHighGrade.test(entity)){
            errs += "Grade must be <= 10!\n";
        }
        if(negativeStId.test(entity)){
            errs += "Student id must be a positive integer!\n";
        }
        if(negativeHwId.test(entity)){
            errs += "Homework id must be a positive integer!\n";
        }
        if(emptyFeedBack.test(entity)){
            errs += "Feedback can't be empty!";
        }
        if(!emptyErrs.test(errs)){
            throw new ValidatorException(errs);
        }
    }
}

File: Semester3/Advanced Programming Methods/Labs/StudentsManagementAppC#/StudentsManagementApp/domain/validator/GradeValidator.cs
using System;
using StudentsManagementApp.domain.entities;

namespace StudentsManagementApp.domain.validator
{
	public class GradeValidator : IValidator<Grade>
	{
		/// <inheritdoc />
		/// <summary>
		/// Function to validate a grade
		/// </summary>
		/// <param name="entity">Grade to be validated</param>
		/// <exception cref="ValidatorException">Throws exception if grade fields are incorrect!</exception>
		public void Validate(Grade entity)
		{
			var errs = "";

			if (entity.GetStudId() < 0)
			{
				errs += "Student id must be a positive integer!\n";
			}
			
			if (entity.GetHwId().Equals(""))
			{
				errs += "Homework Id can't be an empty string!\n";
			}

			if (entity.GradeValue < 0.0 || entity.GradeValue > 10.0)
			{
				errs += "Grade value must be a real number between 0 and 10\n";
			}

			if (entity.Professor.Equals(""))
			{
				errs += "Professor field can't be empty!\n";

			}

			if (entity.FeedBack.Equals(""))
			{
				errs += "Feedback field can't be empty!\n";
			}

			if (!string.IsNullOrEmpty(errs))
			{
				throw new ValidatorException(errs);
			}
		}
	}
}